[MRG+1] ENH Improving the contribution guidelines

NelleV opened the pull request:
Here is a first attempt at improving our documentation on how to contribute. It is greatly inspired from the scikit-learn. I tried to merge all documents into one main document for new contributors.
- created a contributing document that explain all the simple steps to contribute to matplotlib
  - how to fork, clone the git repository.
  - how to install from source
  - how to run the tests
  - how to create a PR and contributions guidelines
- coding_guide has been split into contribution guidelines (in the documentation aforementionned) and reviewers guidelines
- Removed transformation.rst and "hid" the license discussion documentation (which IMO should be removed and replace by a link to Jake V.'s blogpost on that matter).

**What is left to be done**
- [x] Improve the stylesheet for notes
- [x] Improve the content of "The Matplotlib Developers’ Guide"
  - [x] Split the content into three parts: contributor, reviewer, releaser.
  -  [x] Reorder the contents so that it is (1) the main contribution guidelines; (2) documenting mpl; (3) testing mpl; (4) More information on the git workflow; (5) the rest?
- [x] add the new "new-contributor-friendly" tag.

**Open questions**
- Should the "color change" file exist? The information in that file should probably be moved to a MEP
- Should we keep the file "documenting matplotlib"? Most of it is out of date or already documented in sphinx.

An overview of the current documentation can be seen here: http://cbio.mines-paristech.fr/~nvaroquaux/tmp/matplotlib/devel/contributing.html

closes #7234

NelleV commented:
hi @choldgraf 
That's the PR I was talking about. Any feedback is welcomed.

NelleV commented:
A recent build of the documentation is here: http://cbio.mines-paristech.fr/~nvaroquaux/tmp/matplotlib/devel/contributing.html

Any feedback is welcomed.

tacaswell commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,413 @@
+ 2. Fork the `project repository
+    <https://github.com/matplotlib/matplotlib>`__: click on the 'Fork' button
+    near the top of the page. This creates a copy of the code under your
+    account scikit-learnon the GitHub server.
```
scikit-learn reference left.

tacaswell commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,413 @@
+      import matplotlib.collections as mcol
+      import matplotlib.patches as mpatches
+
+  * If your changes are non-trivial, please make an entry in the
```
We do not maintain the CHANGELOG anymore

tacaswell commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,413 @@
+    :file:`CHANGELOG`.
+
+  * If your change is a major new feature, add an entry to
+    :file:`doc/users/whats_new.rst`.
```
ask to create a file in `doc/user/what_new` (and see the README in that folder.  This greatly reduces merge conflicts.  The files are all merged periodically.

tacaswell commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,413 @@
+  * If your change is a major new feature, add an entry to
+    :file:`doc/users/whats_new.rst`.
+
+  * If you change the API in a backward-incompatible way, please
```
ditto for putting in the `doc/api/api_changes` folder.

tacaswell commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,413 @@
+    document it in :file:`doc/api/api_changes.rst`.
+
+
+  * Adding a new pyplot function involves generating code.  See
```
lets not extend the pyplot inteface....

tacaswell commented on doc/devel/reviewers_guide.rst:
```diff
@@ -0,0 +1,112 @@
+* In general, simple bugfixes that are unlikely to introduce new bugs
+  of their own should be merged onto the maintenance branch.  New
+  features, or anything that changes the API, should be made against
+  master.  The rules are fuzzy here -- when in doubt, try to get some
```
I would say 'when in doubt, target master'

tacaswell commented:
I am 50/50 on renaming the rst files.  The new names are better, but they are going to break _somebodys_ bookmark.

tacaswell commented on doc/devel/index.rst:
```diff
@@ -12,14 +12,18 @@ The Matplotlib Developers' Guide
 
-   coding_guide.rst
-   portable_code.rst
-   license.rst
```
I would leave this.

NelleV replied:
Are you talking about the license.rst? It doesn't really fit in the developers section. There is currently a link to it somewhere in the user documentation.

NelleV replied:
Yes, it is

NelleV commented:
@tacaswell good point about the bookmarks. I hadn't thought about this. I'll revert back to the old names.

NelleV commented:
I've addressed @tacaswell's comment (a refresher was clearer a good idea!)
and rebased onto master (thus integrating the PR on setting up matplotlib).

I still have to add the new contributor tag.

NelleV commented:
I've added a section about the new-contibutor-friendly tag and the easy-fix tag and made some changes to the style.

I am still not convinced that the new-contributor-friendly tag adds more to the table than the easy-fix one. we'll have to see long run whether it is worth having both.

A fresh build of the documentation is available here:
http://cbio.mines-paristech.fr/~nvaroquaux/tmp/matplotlib/devel/contributing.html

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+
+ 4. Create a branch to hold your changes::
+
+        $ git checkout -b my-feature
```
I would say "Create a branch, normally from master..." and then append "master" to the checkout command.  I realize it is not necessary for someone following exactly these instructions, but I think it clarifies what is being done.  It gives the new user a hint as to what the line below referring to "master" might mean.

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+    and start making changes. Never work in the ``master`` branch!
+
+ 5. Work on this copy, on your computer, using Git to do the version
+    control. When you're done editing, do::
```
"editing," -> "editing, e.g., 'lib/matplotlib/collections.py'" and then put the example file name in place of "modified_files" below.

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+
+        $ git push -u origin my-feature
+
+Finally, go to the web page of the your fork of the matplotlib repo,
```
delete first "the"

QuLogic replied:
I think you mean the second "the".

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+
+  * If your pull request addresses an issue, please use the title todescribe
+    the issue and mention the issue number in the pull request description
+    to ensure a link is created to the original issue.
```
I think the issue number has to be in the commit message, not the PR description.

QuLogic replied:
It needs to be in the commit message or the description if you want to trigger GitHub's automatic issue closing (which only works when the PR goes to `master` _or_ when the commit (with the reference) is merged into `master`). But cross-references still work everywhere.

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+It is recommended to check that your contribution complies with the following
+rules before submitting a pull request:
+
+  * If your pull request addresses an issue, please use the title todescribe
```
add space -> "to describe"

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+    to ensure a link is created to the original issue.
+
+  * All public methods should have informative docstrings with sample
+    usage presented as doctests when appropriate. Use the
```
Are we actually using doctests anywhere?

NelleV replied:
eh… maybe not. I was actually wondering yesterady if the docstrings were tested, or even just simply run to verify the code actually runs.

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+    usage presented as doctests when appropriate. Use the
+    `numpy docstring standard <https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt>`_
+
+  * Formatting should follow `PEP8 recommandation
```
sp a-> e : recommendation

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+
+    The current state of the matplotlib code base is not compliant with all
+    of those guidelines, but we expect that enforcing those constraints on all
+    new contributions will get the overall code base quality in the right
```
"get" -> "move"

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+
+.. topic:: new-contributor-friendly
+
+    A great way to start contributing to matplotlib is to pick an item from
```
Can this note be consolidated with the block above?  There is quite a bit of duplication here.

NelleV replied:
yeah… I think it would be good to have topics if I had the courage to describe more of the labels that we use on the documentation, but that's not going to happen this week. So I'll just merge everything into one paragraph. 

I had original planned a second topic on easy-fix, but I was rewriting the same thing as for new contributors.

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+-----------------------------------
+
+* If you have added new files or directories, or reorganized existing
+  ones, make sure the new files included in the match patterns in
```
"files included" -> "files are included"

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+  address C++, but most of its admonitions still apply).
+
+* Interfacing with Python may be done either with the raw Python/C API
+  or Cython.
```
We don't have any cython in core matplotlib at present, so I don't think it should be mentioned here until a careful decision has been made to start using it.

NelleV replied:
I removed this line for now. (All of this was just moved from the reviewers's guidelines to here, so there are a lot of elements that could be improved/removed).

efiring replied:
It might need to go back in eventually; I didn't realize it was just moved from elsewhere in our docs.  That suggests that the inclusion of cython was deliberate.  I know we talked about it, but I didn't remember having reached a conclusion.

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+  code.  The interface code should be named `FOO_wrap.cpp` or
+  `FOO_wrapper.cpp`.
+
+* Header file documentation (aka docstrings) should be in Numpydoc
```
I don't understand this bullet point at all.

NelleV replied:
eh… me neither…
I could just remove it?

ivanov replied:
Presumably this is about docstrings that live inside C/C++ code... Like [this `resample` docstring](https://github.com/matplotlib/matplotlib/blob/master/src/_image_wrapper.cpp#L16-L57). I would just change it to "Docstrings should be in Numpydoc..." - since this is already under the `C/C++ extensions` header.

tacaswell replied:
I am in favor of leaving this section as is (but maybe burying it a but further down the page).

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+
+Matplotlib makes extensive use of ``**kwargs`` for pass-through
+customizations from one function to another.  A typical example is in
+:func:`matplotlib.pylab.text`.  The definition of the pylab text
```
pylab -> pyplot

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+i.e., it just passes all ``args`` and ``kwargs`` on to
+:meth:`matplotlib.text.Text.__init__`::
+
+  # in axes.py
```
axes/_axes.py

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+local arguments and the rest are passed on as
+:meth:`~matplotlib.lines.Line2D` keyword arguments::
+
+  # in axes.py
```
axes/_axes.py

efiring commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+:file:`matplotlibrc` (:ref:`customizing-matplotlib`) supports an
+external backend via the ``module`` directive.  if
+:file:`my_backend.py` is a matplotlib backend in your
+:envvar:`PYTHONPATH`, you can set use it on one of several ways
```
delete "use"?  or delete "set"?
All of the following looks like it might be replaced with a link to somewhere else in the docs.

NelleV commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+
+
+After obtaining a local copy of the matpotlib source code (:ref:`set-up-fork`),
+navigate to the matplotlib directory and run the following in the shell:
```
Yes, it is definitely worth it.
I think it might be interesting to have a FAQ of common encountered problems with setting up working environments with matplotlib. This is one of the problem my students encountered, and the other problem being the tests massively failing.

NelleV commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+symbolic links back to the source code). You can then run the tests your work
+environment is set up properly::
+
+  python tests.py
```
On my computer, it works fine at this stage. I am lucky?

QuLogic replied:
It's not strictly problematic, but if you don't tell it to use the fixed version of freetype, you might run into a lot of image failures if all the text has shifted a bit (like say, the next versions of Debian and Fedora, which include a newer FreeType.)

efiring commented:
On 2016/10/12 8:12 AM, Nelle Varoquaux wrote:
> On my computer, it works fine at this stage. I am lucky?

I think it means you just happen to have the right version of freetype 
installed.

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+In case you experience issues using this package, do not hesitate to submit a
+ticket to the
+`Bug Tracker <https://github.com/matplotlib/matplotlib/issues>`_. You are
+also welcome to post feature requests or pull requests.
```
Especially if we continue with "experience issues", then it's even more prudent to mention it.

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+    git clone git@github.com:matplotlib/matplotlib.git
+
+
+After obtaining a local copy of the matpotlib source code (:ref:`set-up-fork`),
```
matpotlib -> matplotlib

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+After obtaining a local copy of the matpotlib source code (:ref:`set-up-fork`),
+navigate to the matplotlib directory and run the following in the shell:
+
+::
```
Is there a reason for not putting the double colons at the end of the previous line?

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+    instance when switching branches or pulling changes from upstream).
+
+.. _nose: http://nose.readthedocs.org/en/latest/
+.. _pep8: http://pep8.readthedocs.org/en/latest/
```
http -> https and .org -> .io for both links.

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+.. note::
+    
+  **Additional dependencies for testing**: nose_ (version 1.0 or later), `mock
+  <http://www.voidspace.org.uk/python/mock/>`_ (if python < 3.3), `Ghostscript
```
This link might be outdated.

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+    
+  **Additional dependencies for testing**: nose_ (version 1.0 or later), `mock
+  <http://www.voidspace.org.uk/python/mock/>`_ (if python < 3.3), `Ghostscript
+  <http://pages.cs.wisc.edu/~ghost/>`_, `Inkscape <http://inkscape.org>`_
```
Ghostscript link might be outdated as well. http -> https on the Inkscape link.

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+
+      * Copy setup.cfg.template to setup.cfg
+      * Edit setup.cfg to set `test` to True, and `local_freetype` to True
+      * If you have build matplotlib previously, remove the build folder.
```
build -> built, for the first one.

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+.. note:: To make sure the tests run locally:
+
+      * Copy setup.cfg.template to setup.cfg
+      * Edit setup.cfg to set `test` to True, and `local_freetype` to True
```
These should be double-backticks, right?

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+    `numpy docstring standard <https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt>`_
+
+  * Formatting should follow `PEP8 recommandation
+    <http://www.python.org/dev/peps/pep-0008/>`_. You should consider
```
http -> https

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+  * Each high-level plotting function should have a simple example in
+    the `Example` section of the docstring.  This should be as simple as
+    possible to demonstrate the method.  More complex examples should go
+    in the `examples` tree.
```
Double backticks on both?
tree -> directory, maybe?

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+      import numpy as np
+      import numpy.ma as ma
+      import matplotlib as mpl
+      from matplotlib import pyplot as plt
```
I think SciPy's convention is `import matplotlib.pyplot as plt`.

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+      import matplotlib as mpl
+      from matplotlib import pyplot as plt
+      import matplotlib.cbook as cbook
+      import matplotlib.collections as mcol
```
There's only one example that uses `mcol`, and most of the rest of the code (no examples, though) uses `mcollections`.

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+      import matplotlib.collections as mcol
+      import matplotlib.patches as mpatches
+
+  * If your change is a major new feature, add an entry to `What's new`
```
to the

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+Note: there is a use case when ``kwargs`` are meant to be used locally
+in the function (not passed on), but you still need the ``**kwargs``
+idiom.  That is when you want to use ``*args`` to allow variable
+numbers of non-keyword args.  In this case, python will not allow you
```
Is that common? It seems like something that's _mostly_ an artifact of Matlab's designs.

tacaswell replied:
And we mirrored that API very closely at the beginning.  There is probably a few more cases (but I do not know any off the top of my head).

This should have a comment added that this is only a python2 restriction, as soon as we drop 2.7 as supported we can use keyword-only arguments :smile: (that is the python3 feature I like the most).

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,422 @@
+----------------
+
+We have hundreds of examples in subdirectories of
+:file:`matplotlib/examples`, and these are automatically generated
```
"these" is ambiguous; it sounds like the _examples_ are automatically generated, but it's really their outputs that are automatically generated.

QuLogic commented on doc/devel/gitwash/development_workflow.rst:
```diff
@@ -1,8 +1,8 @@
-====================
+=========================
+Git Development workflow
+=========================
```
Looks like an extra `=`?

QuLogic commented on doc/devel/portable_code.rst:
```diff
@@ -1,5 +1,9 @@
+.. _portable_code:
+
+=====================================================
+Developer's tips for writing code for python 2 and 3
```
Python

QuLogic commented on doc/devel/testing.rst:
```diff
@@ -17,20 +18,16 @@ Requirements
 
   - nose_, version 1.0 or later
-
   - `mock <http://www.voidspace.org.uk/python/mock/>`_, when running python
```
See above comments, but these links might be outdated.

NelleV commented on doc/_static/mpl.css:
```diff
@@ -366,19 +379,57 @@ div.admonition pre, div.warning pre {
+    margin-top: 10px;
+    padding: 7px;
+    border-radius: 4px;
+    -moz-border-radius: 4px;
```
It's from the nature stylesheet of sphinx.

dopplershift commented on doc/devel/contributing.rst:
```diff
@@ -30,52 +30,56 @@ You can check out the latest sources with the command::
+
+
+.. _nose: https://nose.readthedocs.org/en/latest/
+.. _pep8: https://pep8.readthedocs.org/en/latest/
```
Since you're updating these URLs, would this be a good time to move to readthedocs.io? (See [here](http://blog.readthedocs.com/securing-subdomains/)

NelleV replied:
What do you mean by this? Move the matplotlib documentation to readthedocs.io?

NelleV replied:
oh… I just understood :)

dopplershift commented on doc/devel/contributing.rst:
```diff
@@ -296,7 +288,7 @@ Keyword argument processing
 Matplotlib makes extensive use of ``**kwargs`` for pass-through
 customizations from one function to another.  A typical example is in
-:func:`matplotlib.pylab.text`.  The definition of the pylab text
+:func:`matplotlib.pyplot.text`.  The definition of the pylab text
```
Would it make sense to change `pylab` to `pyplot` in the text as well?

dopplershift commented on doc/devel/testing.rst:
```diff
@@ -10,17 +10,17 @@ customizations to the nose testing infrastructure are in
 please ignore it while we consolidate our testing to these locations.)
 
-.. _nose: http://nose.readthedocs.org/en/latest/
+.. _nose: https://nose.readthedocs.org/en/latest/
```
Again about RTD.

NelleV commented:
@dopplershift I `git grepped` the code for readthedocs.org urls to replace them with the new io ones. The matplotlib documentation should be `rtd.org` free.

QuLogic commented on examples/axisartist/demo_parasite_axes2.py:
```diff
@@ -26,7 +26,7 @@
     par2.set_ylabel("Velocity")
 
-    p1, = host.plot([0, 1, 2], [0, 1, 2], label="Density")
+    p1, = par1.plot([0, 1, 2], [0, 1, 2], label="Density")
```
Was that supposed to change in this commit?

NelleV replied:
that was a mistake. I'll revert it.

dopplershift commented:
Sweet, thanks for going ahead with that!

NelleV commented:
I think this is ready to be merged.

NelleV commented:
@ivanov was interested in giving feedback on this pull request.
I'll rebase in the mean time.

ivanov commented:
👍 -  I wasn't sure why readthedocs links were changed from .org to point to .io, so I looked it up and found [this explanation](https://blog.readthedocs.com/securing-subdomains/) (in case anyone else ends up wondering about that the way I did)

NelleV commented:
Is there anything else to be done on this PR for it to be merged?

QuLogic commented:
Now that GitHub has restored outdated diffs, I can see a few comments above that maybe haven't been handled?

NelleV commented on doc/devel/transformations.rst:
```diff
@@ -1,20 +0,0 @@
-==============================
```
It seems not… I'll add to the the API doc.

tacaswell commented on doc/devel/coding_guide.rst:
```diff
@@ -182,6 +61,12 @@ PR Review guidelines
   'approve review').  If you do the merge please removed the
   ``'[MRG+N']`` prefix.
 
+* Make sure the Travis tests are passing before merging.
```
'travis and appveyor'?

tacaswell commented on doc/devel/coding_guide.rst:
```diff
@@ -182,6 +61,12 @@ PR Review guidelines
+
+  - The Travis tests automatically test on all of the Python versions
+    matplotlib supports whenever a pull request is created or updated.
+    The `tox` support in matplotlib may be useful for testing locally.
```
does the tox setup actually work?

NelleV replied:
Not entirely sure… I think this is some copy paste from the old contribution guidelines.
I think at some point we should go through this document again to make sure everything is up to date.

tacaswell commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,414 @@
+Submitting a bug report
+=======================
+
+If you find a bug in the code or documentation, do not hesitate to submit a
```
Add a note pointing usage questions to gitter or mailing list?

tacaswell commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,414 @@
+
+.. note:: To make sure the tests run locally:
+
+      * Copy setup.cfg.template to setup.cfg
```
I think the env approach here is better, it will survive `git clean -xfd`.  `export MPLLOCALFREETYPE=1` is in my bashrc.

NelleV replied:
I am not sure what you are referring too.
I actually never had to do this to have the tests run on my computer.

tacaswell replied:
If that env is set to anything 'true-ish' during `python setup.py build` it has the same effect as setting this value in setup.cfg and we will pull down the correct freetype, build it, and then link against our version instead of what ever is available on the system.

NelleV commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,414 @@
+
+You can then run the tests to check your work environment is set up properly::
+
+  python tests.py
```
It would be nice to have a shortcut for --processes to -j, as this is quite standard.

tacaswell commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,414 @@
+
+.. _other_ways_to_contribute:
+
+Other ways to contribute
```
I suggest naming this section 'Contributing to documentation' and putting it first.   The tex laments that documentation is under valued while putting it under a title that makes it sound under valued :stuck_out_tongue_winking_eye: 

At risk of being overly micro-managing in comments, something to the effect of

----

We ❤️ documentation improvements and improving the documentation can be one of the highest-impact contributions.  If you can _use_ matplotlib you are qualified to contribute to the documentation.  Did you find the docstring on a confusing?  Fix it so the next person has an easier time.  Could an [link to gallery]Example be improved?  Open a PR!  

----

I blame @rgbkrk for me urge to use emoji in 'serious' text.

rgbkrk replied:
¯\\\_(ツ)\_/¯

rgbkrk replied:
It can help in setting the tone for sections like this, I certainly get it.

tacaswell commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,414 @@
+
+Matplotlib makes extensive use of ``**kwargs`` for pass-through
+customizations from one function to another.  A typical example is in
+:func:`matplotlib.pyplot.text`.  The definition of the pylab text
```
pylab -> pyplot

tacaswell commented on doc/devel/contributing.rst:
```diff
@@ -0,0 +1,414 @@
+function is a simple pass-through to
+:meth:`matplotlib.axes.Axes.text`::
+
+  # in pylab.py
```
`pyplot.py`

tacaswell commented on doc/devel/coding_guide.rst:
```diff
@@ -224,165 +109,3 @@ the merge notification) or through the git CLI tools.::
-
-.. _new-pyplot-function:
-
-Writing a new pyplot function
```
:+1: on dropping this section.

tacaswell commented on doc/devel/coding_guide.rst:
```diff
@@ -1,103 +1,31 @@
-
-  Docstrings should look like (at a minimum)::
-
-        def foo(bar, baz=None):
```
I think keeping a minimal example around is a good idea.  Most people are doing to do this by pattern matching not by reading the spec :smiling_imp: .

tacaswell commented:
This PR now has over 100 comments, this should be merged sooner rather than later.

I left a bunch of comments, but none of them are worth holding this PR up over.

choldgraf commented:
FWIW, I'm trying to get up-to-speed on how to contribute to matplotlib, and this PR would be helpful for somebody in my position. It looks like this is close to being ready to go, looking forward to seeing some updated docs

ivanov commented:
There's one failure on Travis, which is unrelated, since this PR does not introduce any code changes. (for reference - the failure's `test_invisible_Line_rendering`)

I'm going to go ahead and merge this one, and we can iterate further. Good work everyone, especially @NelleV !

Happy hacking! :bowtie:

choldgraf commented:
🍻

ivanov commented:
backported to `v2.x` in 5ac0881f770c4cb0f67d0625b9762264a3cbdfa2

ivanov commented:
(oops, had the wrong SHA in github comment above, fixed it now - it's 28f4b6e6a623c3dfbdd2dcc0df220fe8726f08a7 )

NelleV commented:
Thanks @ivanov ! And welcome back :)
